keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1301 @@
+  feature gates, configs and flags are enabled to match that of a previous
+  release version.
+
+Adding `--emulation-version` to kubelet is out of scope for this enhancement.
+But we do need to define how kubelet skew behaves when the kube-apiserver has
+`--emulation-version` set. Our general rule is that we want to define skew
+using emulation versions when they are in use. So not only must a kubelet
+version be <= the kube-apiserver binary version, it must also be <= the
+`--emulation-version` of the kube-apiserver.
+
+#### --min-compatibility-version
+
+- Defaults to `binaryVersion-1` (matching current behavior)
+- Must be <= `--emulation-version`
+- Must not be lower than the supported range of minor versions (see graduation
+  criteria for ranges). If below the supported version range the binary will
+  exit and report an invalid flag value error telling the user what versions are
+  allowed.
+
+### Changes to Feature Gates
+
+Features will track version information, i.e.:
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Alpha},
+		},
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("1.26"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.28"), Default: true, PreRelease: Deprecated},
+		}
+```
+
+Features with compatibility implications (like a new API field or relaxing
+validation to allow a new enum value) could include that in their feature-gate
+spec:
+
+```go
+map[Feature]VersionedSpecs{
+		featureA: VersionedSpecs{
+			  {Version: mustParseVersion("1.28"), MinCompatibilityVersion: mustParseVersion("1.28"), ...},Alpha},
+			  {Version: mustParseVersion("1.29"), MinCompatibilityVersion: mustParseVersion("1.28"), ...},Beta},
+			  {Version: mustParseVersion("1.30"), MinCompatibilityVersion: mustParseVersion("1.28"), ...},
+		},
+```
+
+And features using a gate to guard a behavior change with compatibility
+implications that isn't really going through the feature lifecycle could set the
+feature version and the min compatibility version to the same thing:
+
+```go
+relaxValidationFeatureA: VersionedSpecs{
+    {Version: mustParseVersion("1.30"), MinCompatibilityVersion: mustParseVersion("1.30"), Default: true, ...},
+},
+```
+
+When a component starts, feature gates `VersionedSpecs` will be compared against
+the emulation and compatibility version to determine which features to enable.
+
+#### Feature Gate Lifecycles
+
+`--feature-gates` must behave the same as it did for the Kubernetes version the
+emulation and compatibility versions are set to. For example, it must be
```

Comment:
Can these conflict? I mean, if emulation and min-compatibility are set to different versions they would.

Shouldn't this just be "emulation version", which is going to control feature gates, where as min-compatibility controls behind the scenes storage and validation features? Or do I not fully get this?

Reply:
You have the core idea.  There is the edge case of feature removals, which impacts compatibility ranges. I'll try to rephrase this.

Reply:
Oh wait, you're right. This actually works out to be quite simple.